README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# fcmsafety

<!-- badges: start -->
<!-- badges: end -->

The goal of fcmsafety is to assign toxicity levels for compounds from
food contact materials. It takes into account the toxicity data from: 1.
Substances of Very High Concern (SVHC) from ECHA
(<https://echa.europa.eu/candidate-list-table>); 2. Carcinogenic,
Mutagenic, and Reprotoxic (CMR) from the Classification, Labelling, and
Packaging (CLP) regulation
(<https://echa.europa.eu/information-on-chemicals/annex-vi-to-clp>) ; 3.
Carcinogenic substances from IARC
(<https://monographs.iarc.who.int/list-of-classifications>); 4.
Endocrine Disrupting Chemicals (EDC) from The International Panel on
Chemical Pollution (IPCP) commissioned by UN Environment
(<https://www.unep.org/explore-topics/chemicals-waste/what-we-do/emerging-issues/scientific-knowledge-endocrine-disrupting>);
5. Specific Migration Limit (SML) from EU 10/2011 regulation
(<https://eur-lex.europa.eu/legal-content/EN/TXT/HTML/?uri=CELEX:02011R0010-20200923&qid=1636402301680&from=en>)
6. Specific Migration Limit (SML) from China GB 9685 regulation
(Provided by IQTC).

The toxicity level assignment follows the rules listed in the
inst/toxicity\_levels.png (don’t know why it is not rendered in this
readme file)

Please bear in mind that the toxicity level assignment here is on the
basis of individual compounds, whilst substances listed in the
aforementioned databases could be groups of compounds. These groups of
substances will not be considered as they not are specific in terms of
chemical structures and no “InChIKey” can be retrieved from Pubchem,
which is used for matching. For example, nonylphenols are a family of
closely related organic compounds composed of phenol bearing a 9
carbon-tail, and they could be overlooked unless any database has
numerated some specific structures in their lists.

In addition, this package provides simple wrappers of some functions
from the **webchem** and **classyfireR** package to ease CID, InChIKey,
SMILES, Formula, ExactMass, CAS, flavonet, and classifire information
retrieval.

## Release notes (2021.12.03 Version 0.1.4)

### Enhancement

1.  Add a function to assign meta data.

## Installation

You can install the development version of fcmsafety from
[GitHub](https://github.com/) with the code listed below. If you have
installed the package before the date listed in *Release Notes*, please
re-install it to ensure to have the latest version.

``` r
install.packages("devtools")
devtools::install_github("QizhiSu/fcmsafety")
```

If you fail to install it due to the classyfireR issue, please install
classyfireR first.

``` r
devtools::install_github('aberHRML/classyfireR')
```

## How to use?

Below walks you through the way to go step by step.

If you do not have the “rio” and “dplyr” package installed, please
install it first. If you do, just skip this step.

``` r
install.packages("rio")
install.packages("dplyr")
```

### Load required packages

``` r
library(fcmsafety)
library(dplyr)
```

### Update and Load databases

You have to first update all the databases and load them into the global
environment.

``` r
# You don't have to update all the databases as we have prepared them (updated 
# on 2021/11/11) and can be used directly. In case you want to have the latest 
# databases (if you find any update of these databases after 2021/11/11), please
# run update_databases(), it might take long time as retrieving meta data for 
# all these databases is time-consuming.
update_databases()

# If you have updated the databases yourself, please run:
load_databases(use_default = FALSE)
# Otherwise
load_databases()
```

### Process your data

1.  Read in your data

Please name your file in English but not Chinese (Chinese letters are
not well supported in some functions). There should be at least one
column in your data containing either chemical names in English, CAS
number, or InChIKey. The program will use any or all of these columns to
retrieve meta data from Pubchem. A good chemical name is always
favorable.

``` r
# Please enter the path of your data, e.g., "D:/my data/mydata.xlsx".
data <- rio::import("D:/my data/mydata.xlsx")
```

2.  Extract cid and meta data

Please specify which column contains CAS number by the “cas\_col”
argument, InChIKey by “inchikey\_col”, and chemical name by the
“name\_col” argument. You can also specify all these arguments. In this
case, it will first use InChIKey, and then CAS and Name. To get
flavornet information, cas = TRUE is required. Depends on the size of
your data, it might take long time.

``` r
# No CAS
data <- data %>% extract_cid(name_col = 1) %>% extract_meta()
# With CAS
data <- data %>% extract_cid(name_col = 1) %>% extract_meta(cas = TRUE)
# With flavornet
data <- data %>% extract_cid(name_col = 1) %>% extract_meta(cas = TRUE, flavonet = TRUE)
```

If you have some compounds that are not present in Pubchem, for example
many oligomers found in food contact materials. there will be no SMILES
retrieved for these compounds using `extract_meta`. In this case, you
can use `assign_meta` function. However, a *.txt file containing Name
and SMILES of these compounds is required. There are two options to
prepare this text file. One is to prepare it manually, the column names
must be Name and SMILES, respectively (case-insensitive). Another one is
to prepare *.MOL files (case-insensitive) of these molecules and extract
SMILES using `combine_mol2sdf()` and `extract_structure()` functions
from the `mspcompiler` package
[mspcompiler](https://github.com/QizhiSu/mspcompiler). Note that the
name in your *.txt file or *.MOL files have to be consistent with the
one you have in your data as Name is used for matching. Assuming you
have all your \*MOL files in the “D:/my data” folder, then you can
follow these steps:

``` r
# If you have not install the mspcompiler package, please install it following
# the instruction in its Github homepage https://github.com/QizhiSu/mspcompiler.
# Once you have it installed
library(mspcompiler)

# This function combines all *.MOL files in yhe providing folder into a single 
# *.sdf file from which will be used to extract SMILES. 
combine_mol2sdf(input = "D:/my data", 
                output = "D:/my data/mydata.sdf",
                use_filename = TRUE)
# The input here is the output from the last command and it will generate a *.txt 
# file containing Name and SMILES.  
extract_structure(input = "D:/my data/mydata.sdf",
                  output = "D:/my data/mydata.txt")
data <- data %>% assign_meta(meta_file = "D:/my data/mydata.txt")
```

3.  Extract classyfire information

After extracting meta data from Pubchem by extract\_meta(), which means
the data will has a column named InChIkey, then we can get the chemical
structure classification done by classyfire.

``` r
data <- data %>% extract_classyfire()
```

4.  Export file for Toxtree

For compounds that do not present in any of the databases we consider
here, we can use Toxtree (<http://toxtree.sourceforge.net/>) to predict
its Cramer rules level which will be used for assigning toxicity level
in the following step. We now can export a file that can be used by
Toxtree for batch processing.

``` r
# This code will generate a file name "for_toxtree.csv"
export4toxtree(data, cas_col = 1, name_col = 2)
# Of course, you can also specify the path and name of the output file
export4toxtree(data, cas_col = 1, name_col = 2, output = "D:/data/mydata_for_toxtree.csv")
```

5.  Toxtree batch processing

    <figure>
    <img src="inst/toxtree_batch_processing1.png" width="500" alt="batch processing 1" /><figcaption aria-hidden="true">batch processing 1</figcaption>
    </figure>

    <figure>
    <img src="inst/toxtree_batch_processing2.png" width="500" alt="batch processing 2" /><figcaption aria-hidden="true">batch processing 2</figcaption>
    </figure>

    <figure>
    <img src="inst/toxtree_batch_processing3.png" width="500" alt="batch processing 3" /><figcaption aria-hidden="true">batch processing 3</figcaption>
    </figure>

Normally, this step is quite fast, but we noticed that polyethylene
glycols might take quite long time.

6.  Assign toxicity

This step requires the results from Toxtree (assuming you have the
result in “D:/data/mydata\_toxtree\_results.csv”).

``` r
data_complete <- data %>% 
  assign_toxicity(toxtree_result = "D:/data/mydata_toxtree_results.csv")
```

7.  Export results

``` r
rio::export(data_complete, "D:/data/mydata_result.xlsx")
```


NEWS.md
# fcmsafety 0.1.4
## Add a function to assign meta data


# fcmsafety 0.1.2

## Enhancement

1. extract_cid() now support any of the following keys, "InChIKey", "CAS", 
or "Name", as well as their combinations. 
2. Add a default value to the cas_col argument in evaluate_compound().


## Bug fixes

1. evaluate_compound() "Error: Argument 1 must have names." error fixed.
2. Remove duplicates results if you have duplicate InChIKey values.
3. Remove warnings when load_databases().

## Others

1. Retrieval CAS number in the extract_meta() function instead of in evaluate_compound().



R/assign_toxicity.R
#' Export input file for toxtree batch processing
#'
#' After \code{extract_cid()} and \code{extract_meta()} for your data, you can
#' use this function to export a *.csv file that can be used by the Toxtree
#' software for batch processing.
#'
#' @param data Your data after the \code{extract_cid()} and \code{extract_meta()}
#' steps
#' @param cas_col The index of column that contains CAS information. CAS number
#' is not mandatory for each compound, if no CAS is available, then chemical name
#' will be used for retreival. However, a column index is still required.
#' @param name_col The index of column that contains chemical name.
#' @param output The output file name end with .csv. The default value is
#' "for_toxtree.csv"
#'
#' @return No return but a *.csv file for Toxtree.
#'
#' @import dplyr
#' @importFrom rio export
#'
#' @export
export4toxtree <-
  function(data, cas_col,
           name_col,
           output = "for_toxtree.csv") {
  data <- data %>%
    mutate(NAME = .[, name_col],
           CAS = .[, cas_col],
           SMILES = SMILES) %>%
    select(NAME, CAS, SMILES) %>%
    filter(!is.na(SMILES))

  rio::export(data, output)
}


#' Assign toxicity levels
#'
#' Assign toxicity levels considering SVHC, EDC, IARC, CMR, EU_10/2011, and China
#' GB 9685
#'
#' @param data Your data containing at least InChIKey
#' @param toxtree_result The *.csv file generated by Toxtree batch processing.
#' The defalut value is "toxtree_results.csv", which means the output files in
#' Toxtree batching processing is "toxtree_results.csv". If other output name is
#' used in Toxtree, please use the corresponding name here.
#'
#' @return A data.frame or tibble with toxicity assigned.
#'
#' @import dplyr
#' @importFrom utils read.csv
#'
#' @export
assign_toxicity <-
  function(data, toxtree_result = "toxtree_results.csv") {
  # read in toxtree result
  tox <- read.csv(toxtree_result)

  data <- data %>%
    mutate(
      Cramer_rules = tox$Cramer.rules[match(SMILES, tox$SMILES)],
      # to check if the compounds present in any of the database
      SVHC = case_when(InChIKey %in% svhc_meta$InChIKey ~ "Y"),
      CMR = case_when(InChIKey %in% cmr_meta$InChIKey ~ "Y"),
      CMR_suspect = case_when(InChIKey %in% cmr_suspect_meta$InChIKey ~ "Y"),
      EDC = case_when(InChIKey %in% edc_meta$InChIKey ~ "Y"),
      IARC = iarc_meta$Group[match(InChIKey, iarc_meta$InChIKey)] %>% na_if("3"),
      EU_SML = eu_sml_meta$SML[match(InChIKey, eu_sml_meta$InChIKey)],
      EU_SML_group = eu_sml_meta$SML_group[match(InChIKey, eu_sml_meta$InChIKey)],
      # it is possible to have both SML and SML_group
      # in this case, use only SML
      EU_SML = case_when(
        is.na(EU_SML) & !is.na(EU_SML_group)
        ~ eu_sml_group$SML[match(EU_SML_group, eu_sml_group$`Group Restriction No`)],
        TRUE ~ EU_SML),
      China_SML = china_sml_meta$SML[match(InChIKey, china_sml_meta$InChIKey)],
      China_SML_group = china_sml_meta$SML_group[match(InChIKey, china_sml_meta$InChIKey)],
      China_SML = case_when(is.na(China_SML & ! is.na(China_SML_group)) ~ China_SML_group,
                            TRUE ~ China_SML),

      Cramer_rules = case_when(
        !is.na(SVHC) | !is.na(CMR) | !is.na(CMR_suspect) | !is.na(EDC)
        | !is.na(IARC) | !is.na(EU_SML) | !is.na(China_SML) ~ "-",
        TRUE ~ Cramer_rules
      ),

      # assign toxiticy
      Toxic_level = case_when(
        !is.na(SVHC) | !is.na(CMR) | !is.na(EDC) | IARC == "1" |
          EU_SML <= 0.018 | China_SML <= 0.018 ~ "V",

        !is.na(CMR_suspect) | IARC == "2A" | IARC == "2B"
        | Cramer_rules == "High (Class III)" |
          (EU_SML > 0.018 & EU_SML <= 0.09) | (China_SML > 0.018 & China_SML <= 0.09) ~ "IV",

        Cramer_rules == "Intermediate (Class II)" |
          (EU_SML > 0.09 & EU_SML <= 0.54) | (China_SML > 0.09 & China_SML <= 0.54) ~ "III",

        Cramer_rules == "Low (Class I)" |
          (EU_SML > 0.54 & EU_SML <= 1.8) | (China_SML > 0.54 & China_SML <= 1.8) ~ "II",

        (EU_SML > 1.8 & EU_SML <= 60) | (China_SML > 1.8 & China_SML <= 60) ~ "I"
      ),

      # mark group SML
      EU_SML = case_when(!is.na(EU_SML_group) ~ paste0(EU_SML, "*"),
                         TRUE ~ as.character(EU_SML)),
      China_SML = case_when(!is.na(China_SML_group) ~ paste0(China_SML, "*"),
                            TRUE ~ as.character(China_SML)),
      EU_SML_group = NULL,
      China_SML_group = NULL
    )

  # replace NA with "-
  data[is.na(data)] <- "-"

  return(data)
}


R/compound_evaluation.R
#' Export input file for toxtree batch processing
#'
#' After \code{extract_cid()} and \code{extract_meta()} for your data, you can
#' use this function to export a *.csv file that can be used by the Toxtree
#' software for batch processing.
#'
#' @param data Your data after the \code{extract_cid()} and \code{extract_meta()}
#' steps
#' @param cas_col The index of column that contains CAS information. CAS number
#' is not mandatory for each compound, if no CAS is available, then chemical name
#' will be used for retreival. However, a column index is still required.
#' @param name_col The index of column that contains chemical name.
#' @param output The output file name end with .csv. The default value is
#' "for_toxtree.csv"
#'
#' @return No return but a *.csv file for Toxtree.
#'
#' @import dplyr
#' @importFrom rio export
#'
#' @export
export4toxtree <-
  function(data, cas_col,
           name_col,
           output = "for_toxtree.csv") {
  data <- data %>%
    mutate(NAME = .[, name_col],
           CAS = .[, cas_col],
           SMILES = SMILES) %>%
    select(NAME, CAS, SMILES) %>%
    filter(!is.na(SMILES))

  rio::export(data, output)
}


#' Evaluate toxicity, get flavornet and classyfire information
#'
#' @param data Your data after the \code{extract_cid()} and \code{extract_meta()}
#'
#' @param toxtree_result The *.csv file generated by Toxtree batch processing.
#' The defalut value is "toxtree_results.csv", which means the output files in
#' Toxtree batching processing is "toxtree_results.csv". If other output name is
#' used in Toxtree, please use the corresponding name here.
#' @param flavornet A logical value for flavornet information retrieval.
#' @param cas_col In case you want to retrieve flavornet information, please
#' specify the index of column that contains CAS number which can be retrieved
#' using \code{extract_meta()} function.
#' @param classyfire A logical value for classyfire information retrieval.
#'
#' @return A data.frame or tibble with toxicity assigned.
#'
#' @import dplyr
#' @import webchem
#' @import classyfireR
#' @import purrr
#' @importFrom utils read.csv
#'
#' @export
evaluate_compound <-
  function(data,
           toxtree_result = "toxtree_results.csv",
           flavornet = TRUE,
           cas_col = NA,
           classyfire = TRUE
           ) {
  # read in toxtree result
  tox <- read.csv(toxtree_result)

  data <- data %>%
    mutate(
      Cramer_rules = tox$Cramer.rules[match(SMILES, tox$SMILES)],
      # to check if the compounds present in any of the database
      SVHC = case_when(InChIKey %in% svhc_meta$InChIKey ~ "Y"),
      CMR = case_when(InChIKey %in% cmr_meta$InChIKey ~ "Y"),
      CMR_suspect = case_when(InChIKey %in% cmr_suspect_meta$InChIKey ~ "Y"),
      EDC = case_when(InChIKey %in% edc_meta$InChIKey ~ "Y"),
      IARC = iarc_meta$Group[match(InChIKey, iarc_meta$InChIKey)] %>% na_if("3"),
      EU_SML = eu_sml_meta$SML[match(InChIKey, eu_sml_meta$InChIKey)],
      EU_SML_group = eu_sml_meta$SML_group[match(InChIKey, eu_sml_meta$InChIKey)],
      # it is possible to have both SML and SML_group
      # in this case, use only SML
      EU_SML = case_when(
        is.na(EU_SML) & !is.na(EU_SML_group)
        ~ eu_sml_group$SML[match(EU_SML_group, eu_sml_group$`Group Restriction No`)],
        TRUE ~ EU_SML),
      China_SML = china_sml_meta$SML[match(InChIKey, china_sml_meta$InChIKey)],
      China_SML_group = china_sml_meta$SML_group[match(InChIKey, china_sml_meta$InChIKey)],
      China_SML = case_when(is.na(China_SML & ! is.na(China_SML_group)) ~ China_SML_group,
                            TRUE ~ China_SML),

      Cramer_rules = case_when(
        !is.na(SVHC) | !is.na(CMR) | !is.na(CMR_suspect) | !is.na(EDC)
        | !is.na(IARC) | !is.na(EU_SML) | !is.na(China_SML) ~ "-",
        TRUE ~ Cramer_rules
      ),

      # assign toxiticy
      Toxic_level = case_when(
        !is.na(SVHC) | !is.na(CMR) | !is.na(EDC) | IARC == "1" |
          EU_SML <= 0.018 | China_SML <= 0.018 ~ "V",

        !is.na(CMR_suspect) | IARC == "2A" | IARC == "2B"
        | Cramer_rules == "High (Class III)" |
          (EU_SML > 0.018 & EU_SML <= 0.09) | (China_SML > 0.018 & China_SML <= 0.09) ~ "IV",

        Cramer_rules == "Intermediate (Class II)" |
          (EU_SML > 0.09 & EU_SML <= 0.54) | (China_SML > 0.09 & China_SML <= 0.54) ~ "III",

        Cramer_rules == "Low (Class I)" |
          (EU_SML > 0.54 & EU_SML <= 1.8) | (China_SML > 0.54 & China_SML <= 1.8) ~ "II",

        (EU_SML > 1.8 & EU_SML <= 60) | (China_SML > 1.8 & China_SML <= 60) ~ "I"
      ),

      # mark group SML
      EU_SML = case_when(!is.na(EU_SML_group) ~ paste0(EU_SML, "*"),
                         TRUE ~ as.character(EU_SML)),
      China_SML = case_when(!is.na(China_SML_group) ~ paste0(China_SML, "*"),
                            TRUE ~ as.character(China_SML)),
      EU_SML_group = NULL,
      China_SML_group = NULL
    )

  if(flavornet == TRUE) {
    data$Flavornet = webchem::fn_percept(data[, cas_col])
    }

  if(classyfire == TRUE) {
    # extract classification from classyfireR
    classyfire <- data$InChIKey %>%
      unique() %>% # remove duplicates
      purrr::map(classyfireR::get_classification) %>%
      purrr::discard(is.null)
    # extract meta data for classification
    classyfire_meta <- classyfire %>%
      purrr::map(classyfireR::meta)%>%
      sapply("[[", 1) %>%
      str_remove("InChIKey=") %>%
      as_tibble() %>%
      rename(InChIKey = value)

    data <-
      classyfire %>%
      purrr::map(classyfireR::classification) %>%
      lapply(extract_cla) %>%
      do.call("bind_rows", .) %>% # collapse the list into a data.frame
      cbind(classyfire_meta, .) %>% # join with InChIKey
      left_join(data, ., by = "InChIKey") # join with the raw table
  }

  # replace NA with "-
  data[is.na(data)] <- "-"

  return(data)
}


R/databases.R
#' Update SVHC, CMR, IARC, and EU_SML database
#'
#' \code{update_databases()} provides a way to update the Substances of Very High
#' Concern (SVHC), Carcinogenic, Mutagenic, and Reprotoxic (CMR) compounds
#' from the Classification, Labelling, and Packagine (CLP) regulation Table 3 of
#' Annex VI, Carcinogenic substances from the International Agency for Research
#' on Cancer (IARC), as well as the positive list of the EU 10/2011 regulation on
#' plastic food contact materials. It then extracts CID and meta data, e.g.,
#' IsomericSMILES, InChIKey. It will first check if there is a \strong{inst} folder
#' in the working directory. If no, it will create one and generate
#' some *.xlsx files that can be read in later on with the \code{load_databases()}
#' function. If yes, it will update existed *.xlsx files. This process can be
#' time-consuming (up to several hours to retrieve information from Pubchem) and
#' It is not necessary to update all databases every time you want to use them
#' because these databases will not be frequently updated by the holders. Moreover,
#' Chrome brower is required and it will prompt up a Chrome windows for few seconds.
#' Please do not close it. It will disappear after downloading the required database.
#'
#' @param svhc A logical value to decide whether or not to update this database.
#' @param cmr A logical value to decide whether or not to update this database.
#' @param iarc A logical value to decide whether or not to update this database.
#' @param eu_sml A logical value to decide whether or not to update this database.
#'
#' @return It will return updated databases with CID and meta data assigned.
#'
#' @export
#'
#' @import webchem
#' @import dplyr
#' @import stringr
#' @import rvest
#' @import httr
#' @import RSelenium
#' @importFrom binman list_versions
#' @importFrom netstat free_port
#' @importFrom fs dir_ls
#' @importFrom janitor row_to_names
#' @importFrom magrittr extract
#' @importFrom utils download.file
update_databases <-
  function(svhc = TRUE, cmr = TRUE, iarc = TRUE, eu_sml = TRUE) {
    if(!dir.exists(paste0(getwd(),"/inst")))
      dir.create(paste0(getwd(),"/inst"))

    if (svhc == TRUE) {
      # SVHC https://echa.europa.eu/candidate-list-table
      # the url is available via checking the Network tab in the Inspect menu in Chrome
      url <-
        'https://echa.europa.eu/candidate-list-table?p_p_id=disslists_WAR_disslistsportlet&p_p_lifecycle=2&p_p_state=normal&p_p_mode=view&p_p_resource_id=exportResults&p_p_cacheability=cacheLevelPage'
      httr::POST(
        url,
        body = list(
          "_disslists_WAR_disslistsportlet_formDate" = as.numeric(as.POSIXct(Sys.time())) * 1000,
          "_disslists_WAR_disslistsportlet_exportColumns" = "name,ecNumber,casNumber,haz_detailed_concern,dte_inclusion,doc_cat_decision,doc_cat_iuclid_dossier,doc_cat_supdoc,doc_cat_rcom,prc_external_remarks",
          "_disslists_WAR_disslistsportlet_orderByCol" = "dte_inclusion",
          "_disslists_WAR_disslistsportlet_orderByType" = "desc",
          "_disslists_WAR_disslistsportlet_searchFormColumns" = "haz_detailed_concern,dte_inclusion",
          "_disslists_WAR_disslistsportlet_searchFormElements" = "DROP_DOWN,DATE_PICKER",
          "_disslists_WAR_disslistsportlet_substance_identifier_field_key" = "",
          "_disslists_WAR_disslistsportlet_haz_detailed_concern" = "",
          "_disslists_WAR_disslistsportlet_dte_inclusionFrom" = "",
          "_disslists_WAR_disslistsportlet_dte_inclusionTo" = "",
          "_disslists_WAR_disslistsportlet_total" = "219",
          "_disslists_WAR_disslistsportlet_exportType" = "xls"
        ),
        httr::write_disk(paste0(getwd(), "/inst/svhc.xlsx"), overwrite = TRUE)
      )

      # read in the downloaded xlsx file
      svhc <- rio::import(paste0(getwd(), "/inst/svhc.xlsx"), skip = 3)
      svhc <- svhc %>%
        # this compound cannot be extract by CAS neither name (weird name)
        mutate(
          `Substance name` =
            str_replace(
              .$`Substance name`,
              "^Trixylyl phosphate$",
              "Trixylenyl phosphate"
            )
        ) %>%
        extract_cid(cas_col = 4, name_col = 1)
      svhc_meta <- extract_meta(svhc)
      # export the table with meta data
      rio::export(svhc_meta, paste0(getwd(), "/inst/svhc.xlsx"))
    }


    if (cmr == TRUE) {
      # extract CMR compounds from CLP regulation Table 3 of Annex VI (latest one)
      url <- "https://echa.europa.eu/information-on-chemicals/annex-vi-to-clp"
      # extract href from all a tags and then fileter on the base of "documents/10162.."
      url_list <- url %>%
        read_html() %>%
        html_nodes("a") %>%
        html_attr("href") %>%
        as_tibble() %>%
        filter(str_detect(value, "documents/10162/17218/annex_vi_clp"))
      # have to add https//echa.europa.eu at the beginning
      url <- paste0("https://echa.europa.eu", url_list[nrow(url_list),])
      # download the latest clp list
      download.file(url, paste0(getwd(), "/inst/clp.xlsx"), quiet = TRUE, mode = "wb")

      # import the download latest clp list file
      clp <- suppressMessages(rio::import(paste0(getwd(), "/inst/clp.xlsx"), skip = 3))
      # rename some columns
      for (i in 1:ncol(clp)) {
        if (!is.na(clp[1, i]))
          colnames(clp)[i] <- clp[1, i]
      }
      clp <- clp[-1, ] # remove first row

      # subset for CMR and suspect CMR
      cmr <- clp %>%
        filter(str_detect(`Hazard Statement Code(s)`, "H340|H350|H360")) %>%
        extract_cid(cas_col = 4, name_col = 2)
      cmr_meta <- extract_meta(cmr)
      cmr_suspect <- clp %>%
        filter(str_detect(`Hazard Statement Code(s)`, "H341|H351|H361")) %>%
        extract_cid(cas_col = 4, name_col = 2)
      cmr_suspect_meta <- extract_meta(cmr_suspect)
      # export cmr and cmr_suspect to a single file but different sheet
      rio::export(list(cmr = cmr_meta, cmr_suspect = cmr_suspect_meta),
                  paste0(getwd(), "/inst/clp_cmr_meta.xlsx"))
    }


    if (iarc == TRUE) {
      # the IARC database, https://monographs.iarc.who.int/agents-classified-by-the-iarc/
      url <- "https://monographs.iarc.who.int/list-of-classifications"
      path <- gsub("/", "\\\\", paste0(getwd(), "/inst")) # set the working directory for downloading
      # set up extra capabilities
      ecaps <- list(chromeOptions = list(
        prefs = list(
          "profile.default_content_settings.popups" = 0L,
          "download.prompt_for_download" = FALSE,
          "download.default_directory" = path
        ),
        args = c('--disable-gpu', '--window-size=600,800')
      ))
      # lauch a chrome driver
      driver <- RSelenium::rsDriver(
        browser = "chrome",
        # the long chromever avoid imcompatible chrome version
        # https://thatdatatho.com/tutorial-web-scraping-rselenium/
        chromever =
          system2(
            command = "wmic",
            args = 'datafile where name="C:\\\\Program Files (x86)\\\\Google\\\\Chrome\\\\Application\\\\chrome.exe" get Version /value',
            stdout = TRUE,
            stderr = TRUE
          ) %>%
          stringr::str_extract(pattern = "(?<=Version=)\\d+\\.\\d+\\.\\d+\\.") %>%
          magrittr::extract(!is.na(.)) %>%
          stringr::str_replace_all(pattern = "\\.",
                                   replacement = "\\\\.") %>%
          paste0("^",  .) %>%
          stringr::str_subset(
            string =
              binman::list_versions(appname = "chromedriver") %>%
              dplyr::last()
          ) %>%
          as.numeric_version() %>%
          max() %>%
          as.character(),
        # use free port to allowed being repeatedly executed
        port = netstat::free_port(),
        extraCapabilities = ecaps
      )

      remote_driver <- driver[["client"]]
      remote_driver$navigate(url)
      buttom_element <-
        remote_driver$findElement(
          using = "xpath",
          value = '//*[@id="table_wrapper"]/div[1]/button[3]')
      buttom_element$clickElement()
      Sys.sleep(2)
      remote_driver$close()

      # get the downloaded file name
      all_files <- fs::dir_ls(paste0(getwd(), "/inst"), glob = "*.xlsx") %>%
        str_remove_all(paste0(getwd(), "/inst/"))
      file_name <- all_files[str_which(
        all_files,
        "Agents Classified by the IARC Monographs")]
      # move the file to the desired directory
      file.rename(file.path(paste0(getwd(), "/inst/"), file_name),
                  file.path(paste0(getwd(), "/inst/iarc.xlsx")))

      # read in the file and extract meta
      iarc <- rio::import(paste0(getwd(), "/inst/iarc.xlsx"), skip = 1)
      iarc <- iarc %>% extract_cid(cas_col = 1, name_col = 2)
      iarc_meta <- extract_meta(iarc)
      rio::export(iarc_meta, paste0(getwd(), "/inst/iarc_meta.xlsx"))
    }


    if (eu_sml == TRUE) {
      # EU 10/2011, no up to date
      url <- "https://eur-lex.europa.eu/legal-content/EN/TXT/HTML/?uri=CELEX:02011R0010-20200923&qid=1636402301680&from=en"
      # nodes of all tables
      eu_nodes <- read_html(url) %>% html_elements("div.centered table tbody")
      eu_sml <-  eu_nodes %>%
        .[[1]] %>%
        html_table(header = TRUE) %>%
        suppressWarnings() %>%
        janitor::row_to_names(row_number = 1) %>%
        filter(!str_detect(`FCM substance No`, "\u25bc|\\(")) %>%
        distinct(`FCM substance No`, .keep_all = TRUE)
      # the group SML
      eu_sml_group <- eu_nodes %>%
        .[[2]] %>%
        html_table(header = TRUE) %>%
        suppressWarnings() %>%
        janitor::row_to_names(row_number = 1) %>%
        filter(!str_detect(`FCM substance No`, "\u25bc")) %>%
        distinct(`Group Restriction No`, .keep_all = TRUE)
      # export the extacted table into a single excel file
      rio::export(list(SML = eu_sml, SML_group = eu_sml_group),
                  paste0(getwd(), "/inst/eu10_2011.xlsx"))

      # clean CAS and extract meta
      eu_sml <- eu_sml %>%
        mutate(`CAS No` = str_remove(.$`CAS No`, "^0*")) %>%
        mutate(`CAS No` = str_remove(.$`CAS No`, "\n.*")) %>%
        as.data.frame() %>% # don't know why as_data_frame or tibble do not work
        extract_cid(cas_col = 3, name_col = 4)
      eu_sml_meta <- extract_meta(eu_sml)
      rio::export(eu_sml_meta, paste0(getwd(), "/inst/eu10_2011_meta.xlsx")) # export complete list
    }
  }


#' Load all databases with meta data
#'
#' \code{load_databases()} provides a way to load all databases with meta data
#' into the global environment. It requires no argument but cleaned up databases
#' files in *.xlsx format in the \strong{inst} folder. Pre-prepared databases
#' (updated on 2021/11/11) already exist in the \strong{inst} folder. If you
#' \code{update_databases()}, theses databases will be updated to the latest one.
#' At the moment, the EDC and China SML (GB 9685) databases can not be updated.
#'
#' @param use_default A logical value. If use_default = TRUE, it will load
#' the pre-prepared databases. Otherwise, it will load the use-updated databases
#' using the \code{update_databases()} function. In the latter case, a \strong{inst}
#' folder containing all the databases (*.xlsx) is required.
#'
#' @return data.frame of all databases.
#'
#' @export
#'
#' @importFrom rio import export
load_databases <- function(use_default = TRUE) {
  if(use_default == FALSE) {
    svhc_meta <- rio::import(paste0(getwd(), "/inst/svhc_meta.xlsx"))
    cmr_meta <- rio::import(paste0(getwd(), "/inst/clp_cmr_meta.xlsx"),
                             sheet = "cmr")
    cmr_suspect_meta <- rio::import(paste0(getwd(), "/inst/clp_cmr_meta.xlsx"),
                                     sheet = "cmr_suspect")
    iarc_meta <- rio::import(paste0(getwd(), "/inst/iarc_meta.xlsx"))
    eu_sml_meta <- rio::import(paste0(getwd(), "/inst/eu10_2011_meta.xlsx"))
    eu_sml_group <- rio::import(paste0(getwd(), "/inst/eu10_2011.xlsx"),
                                 sheet = "SML_group")
    edc_meta <- rio::import(paste0(getwd(), "/inst/edc_meta.xlsx"))
    china_sml_meta <- rio::import(paste0(getwd(), "/inst/china_sml_meta_cleaned.xlsx"))
  } else {
    svhc_meta <- rio::import(system.file("svhc_meta.xlsx", package = "fcmsafety"))
    cmr_meta <- rio::import(system.file("clp_cmr_meta.xlsx", package = "fcmsafety"),
                             sheet = "cmr")
    cmr_suspect_meta <- rio::import(system.file("clp_cmr_meta.xlsx",
                                                 package = "fcmsafety"),
                                     sheet = "cmr_suspect")
    iarc_meta <- rio::import(system.file("iarc_meta.xlsx", package = "fcmsafety"))
    eu_sml_meta <- rio::import(system.file("eu10_2011_meta.xlsx", package = "fcmsafety"))
    eu_sml_group <- rio::import(system.file("eu10_2011.xlsx", package = "fcmsafety"),
                                 sheet = "SML_group")
    edc_meta <- rio::import(system.file("edc_meta.xlsx", package = "fcmsafety"))
    china_sml_meta <- rio::import(system.file("china_sml_meta_cleaned.xlsx",
                                               package = "fcmsafety")) %>%
      suppressWarnings()
  }

  # read in the glogal environment and clean them up
  svhc_meta <<- svhc_meta %>% filter(!is.na(InChIKey))
  cmr_meta <<- cmr_meta %>% filter(!is.na(InChIKey))
  cmr_suspect_meta <<- cmr_suspect_meta %>% filter(!is.na(InChIKey))
  iarc_meta <<- iarc_meta %>% filter(!is.na(InChIKey))
  eu_sml_meta <<- eu_sml_meta %>%
    filter(!is.na(InChIKey)) %>%
    rename(SML = `SML\r\n                     [mg/kg]`,
           SML_group = `SML(T)\r\n                     [mg/kg]\r\n                     (Group restriction No)`) %>%
    mutate(SML = SML %>%
             str_replace(",", ".") %>%
             str_replace("ND", "0.01") %>%
             # soybean oil expoxidized have 2 SML, keep only the first one
             str_remove_all("\n.*$") %>%
             trimws() %>%
             as.numeric(),
           SML_group = str_remove_all(SML_group, "\\(|\\)"))
  eu_sml_group <<- eu_sml_group %>%
    rename(SML = `SML (T)\r\n                     [mg/kg]`) %>%
    mutate(SML = SML %>%
             str_replace(",", ".") %>%
             str_replace("ND", "0.01") %>%
             trimws() %>%
             as.numeric())
  edc_meta <<- edc_meta %>% filter(!is.na(InChIKey))
  china_sml_meta <<- china_sml_meta

}

